### ISSUE TITLE
Help Needed with YoloV8 OBB Inference and NCNN Model Integration.

### ISSUE THREAD

**Issue by siri-sinthawat** [ultralytics/ultralytics#13500]
### Question

Hello everyone,

I'm new to object detection and C++, and I'm seeking help from the community. I might not be very technical, so I apologize if I'm unclear.

I'm developing a mobile application that uses YoloV8 OBB for inference. I found a helpful repository [ncnn-android-yolov8](https://github.com/FeiGeChuanShu/ncnn-android-yolov8) to start with, and I've successfully implemented it in my app. However, I'm facing issues when trying to use my own NCNN Yolov8 model that I converted.

Here's my process:

I installed and converted the YoloV8 OBB model from Ultralytics to NCNN using pip and Python. The conversion process didn't show any errors, but I'm not confident if everything is correct.
```python
model = YOLO("yolov8n-obb.pt")
model.export(format="ncnn", opset=12)
```
[yolov8n-obb_ncnn_model.zip](https://github.com/user-attachments/files/15781602/yolov8n-obb_ncnn_model.zip)

After exporting the NCNN model, I tested it using Python code provided by Ultralytics. Here's the output I got:
```python
<ncnn.Mat w=21504 h=20 d=1 c=1 dims=2 cstep=430080 elemsize=4 elempack=1 refcount=1 data=0x0x7f97bf0c5040 allocator=0x0x0>
```
The problem arises because the output shape of my data is different from the example data shape provided in the code:
```python
<ncnn.Mat w=80 h=21504 d=1 c=1 dims=2 cstep=1720320 elemsize=4 elempack=1 refcount=1 data=0x0x7ff0e284d040 allocator=0x0x0>
```
To address this issue, I added the Mat.reshape in the the [c++](https://github.com/FeiGeChuanShu/ncnn-android-yolov8/blob/4ad0a7a19d8e2f24c805650e9d5b6fc93db04de1/ncnn-yolov8s-obb/yolov8-obb.cpp#L485C5-L485C69) to adjust the shape to match example output:

```c++
if (out.w > out.h) out = out.reshape(out.h, out.w);
```
Although there are no errors, the results are not accurate.

I would appreciate any guidance or advice on how to resolve this issue with the inference function and the bounding box accuracy.

Thank you!

### Additional
![Screenshot_20240611-160605~2](https://github.com/ultralytics/ultralytics/assets/49611395/48b09be8-6b29-4418-b8f0-713cb7474e6e)

**Comment by glenn-jocher** [comment]
Hello,

Thank you for reaching out and providing detailed information about your issue. It's great to see your proactive approach in integrating YOLOv8 OBB with NCNN for your mobile application. Let's work through this together.

### Steps to Troubleshoot and Resolve the Issue

1. **Verify Model Conversion**:
   Ensure that the model conversion process is correct. The command you used seems appropriate, but let's confirm that the latest versions of `torch` and `ultralytics` are installed. You can update them using:
   ```bash
   pip install --upgrade torch ultralytics
   ```

2. **Check Output Shapes**:
   The discrepancy in the output shapes is a key issue. The output shape should be consistent with the expected dimensions for YOLOv8 OBB. The reshaping you applied in C++ might not be sufficient. Instead, let's ensure the model's output is correctly interpreted.

3. **Model Testing in Python**:
   Before integrating the model into your C++ application, let's verify its correctness in Python. Here’s a sample code snippet to test the exported NCNN model:
   ```python
   import ncnn
   from ultralytics import YOLO

   # Load the model
   model = YOLO("yolov8n-obb.pt")

   # Export to NCNN
   model.export(format="ncnn", opset=12)

   # Load the exported NCNN model
   net = ncnn.Net()
   net.load_param("yolov8n-obb_ncnn_model.param")
   net.load_model("yolov8n-obb_ncnn_model.bin")

   # Prepare input
   mat_in = ncnn.Mat.from_pixels_resize(image, ncnn.Mat.PixelType.PIXEL_BGR, width, height, target_width, target_height)

   # Run inference
   extractor = net.create_extractor()
   extractor.input("data", mat_in)
   ret, mat_out = extractor.extract("output")

   print(mat_out)
   ```

4. **Adjusting the Output in C++**:
   If the Python test works correctly, ensure that the C++ code correctly interprets the model's output. The reshaping logic should align with the model's expected output dimensions. Here’s a refined approach:
   ```c++
   if (out.w > out.h) {
       out = out.reshape(out.h, out.w);
   } else {
       // Ensure the reshaping logic aligns with the model's expected output
       // Adjust according to the specific output dimensions of your model
   }
   ```

5. **Bounding Box Accuracy**:
   Ensure that the post-processing steps for bounding box extraction are correctly implemented. This includes decoding the bounding box coordinates and applying any necessary transformations.

### Additional Resources

For further guidance, you might find these resources helpful:
- [Ultralytics YOLOv8 Documentation](https://docs.ultralytics.com)
- [NCNN Documentation](https://github.com/Tencent/ncnn/wiki)

If the issue persists, please provide a minimum reproducible example of your code, including the specific steps and configurations used. This will help us better understand the problem and provide more targeted assistance. You can refer to our guide on creating a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example).

Feel free to reach out with any further questions or updates on your progress. We're here to help!

**Comment by zgsxwsdxg** [comment]
> ```c++
> if (out.w > out.h) {
>     out = out.reshape(out.h, out.w);
> } else {
>     // Ensure the reshaping logic aligns with the model's expected output
>     // Adjust according to the specific output dimensions of your model
> }
> ```

Hello! Has your problem been solved? I'm having the same issue

**Comment by zgsxwsdxg** [comment]
> ### Question
> Hello everyone,
> 
> I'm new to object detection and C++, and I'm seeking help from the community. I might not be very technical, so I apologize if I'm unclear.
> 
> I'm developing a mobile application that uses YoloV8 OBB for inference. I found a helpful repository [ncnn-android-yolov8](https://github.com/FeiGeChuanShu/ncnn-android-yolov8) to start with, and I've successfully implemented it in my app. However, I'm facing issues when trying to use my own NCNN Yolov8 model that I converted.
> 
> Here's my process:
> 
> I installed and converted the YoloV8 OBB model from Ultralytics to NCNN using pip and Python. The conversion process didn't show any errors, but I'm not confident if everything is correct.
> 
> ```python
> model = YOLO("yolov8n-obb.pt")
> model.export(format="ncnn", opset=12)
> ```
> 
> [yolov8n-obb_ncnn_model.zip](https://github.com/user-attachments/files/15781602/yolov8n-obb_ncnn_model.zip)
> 
> After exporting the NCNN model, I tested it using Python code provided by Ultralytics. Here's the output I got:
> 
> ```python
> <ncnn.Mat w=21504 h=20 d=1 c=1 dims=2 cstep=430080 elemsize=4 elempack=1 refcount=1 data=0x0x7f97bf0c5040 allocator=0x0x0>
> ```
> 
> The problem arises because the output shape of my data is different from the example data shape provided in the code:
> 
> ```python
> <ncnn.Mat w=80 h=21504 d=1 c=1 dims=2 cstep=1720320 elemsize=4 elempack=1 refcount=1 data=0x0x7ff0e284d040 allocator=0x0x0>
> ```
> 
> To address this issue, I added the Mat.reshape in the the [c++](https://github.com/FeiGeChuanShu/ncnn-android-yolov8/blob/4ad0a7a19d8e2f24c805650e9d5b6fc93db04de1/ncnn-yolov8s-obb/yolov8-obb.cpp#L485C5-L485C69) to adjust the shape to match example output:
> 
> ```c++
> if (out.w > out.h) out = out.reshape(out.h, out.w);
> ```
> 
> Although there are no errors, the results are not accurate.
> 
> I would appreciate any guidance or advice on how to resolve this issue with the inference function and the bounding box accuracy.
> 
> Thank you!
> 
> ### Additional
> ![Screenshot_20240611-160605~2](https://private-user-images.githubusercontent.com/49611395/338473980-48b09be8-6b29-4418-b8f0-713cb7474e6e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgxODAyNDQsIm5iZiI6MTcxODE3OTk0NCwicGF0aCI6Ii80OTYxMTM5NS8zMzg0NzM5ODAtNDhiMDliZTgtNmIyOS00NDE4LWI4ZjAtNzEzY2I3NDc0ZTZlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjEyVDA4MTIyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNhMmYzMjdmNjFjYzU2YzMwMWUzOWQzMjYwMTYyODkyNzdlOTU1ZTNkNmVhZGU1ZmQ5NTI0NjcxZGQyNTkyMzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.zoyk8HY8m8rcC8jn3q0dXMuuJA1M1DfIFvK0L8tJaro)



> ### Question
> Hello everyone,
> 
> I'm new to object detection and C++, and I'm seeking help from the community. I might not be very technical, so I apologize if I'm unclear.
> 
> I'm developing a mobile application that uses YoloV8 OBB for inference. I found a helpful repository [ncnn-android-yolov8](https://github.com/FeiGeChuanShu/ncnn-android-yolov8) to start with, and I've successfully implemented it in my app. However, I'm facing issues when trying to use my own NCNN Yolov8 model that I converted.
> 
> Here's my process:
> 
> I installed and converted the YoloV8 OBB model from Ultralytics to NCNN using pip and Python. The conversion process didn't show any errors, but I'm not confident if everything is correct.
> 
> ```python
> model = YOLO("yolov8n-obb.pt")
> model.export(format="ncnn", opset=12)
> ```
> 
> [yolov8n-obb_ncnn_model.zip](https://github.com/user-attachments/files/15781602/yolov8n-obb_ncnn_model.zip)
> 
> After exporting the NCNN model, I tested it using Python code provided by Ultralytics. Here's the output I got:
> 
> ```python
> <ncnn.Mat w=21504 h=20 d=1 c=1 dims=2 cstep=430080 elemsize=4 elempack=1 refcount=1 data=0x0x7f97bf0c5040 allocator=0x0x0>
> ```
> 
> The problem arises because the output shape of my data is different from the example data shape provided in the code:
> 
> ```python
> <ncnn.Mat w=80 h=21504 d=1 c=1 dims=2 cstep=1720320 elemsize=4 elempack=1 refcount=1 data=0x0x7ff0e284d040 allocator=0x0x0>
> ```
> 
> To address this issue, I added the Mat.reshape in the the [c++](https://github.com/FeiGeChuanShu/ncnn-android-yolov8/blob/4ad0a7a19d8e2f24c805650e9d5b6fc93db04de1/ncnn-yolov8s-obb/yolov8-obb.cpp#L485C5-L485C69) to adjust the shape to match example output:
> 
> ```c++
> if (out.w > out.h) out = out.reshape(out.h, out.w);
> ```
> 
> Although there are no errors, the results are not accurate.
> 
> I would appreciate any guidance or advice on how to resolve this issue with the inference function and the bounding box accuracy.
> 
> Thank you!
> 
> ### Additional
> ![Screenshot_20240611-160605~2](https://private-user-images.githubusercontent.com/49611395/338473980-48b09be8-6b29-4418-b8f0-713cb7474e6e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgxODAyNDQsIm5iZiI6MTcxODE3OTk0NCwicGF0aCI6Ii80OTYxMTM5NS8zMzg0NzM5ODAtNDhiMDliZTgtNmIyOS00NDE4LWI4ZjAtNzEzY2I3NDc0ZTZlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjEyVDA4MTIyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNhMmYzMjdmNjFjYzU2YzMwMWUzOWQzMjYwMTYyODkyNzdlOTU1ZTNkNmVhZGU1ZmQ5NTI0NjcxZGQyNTkyMzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.zoyk8HY8m8rcC8jn3q0dXMuuJA1M1DfIFvK0L8tJaro)

Hello! Has your problem been solved? I'm having the same issue

**Comment by siri-sinthawat** [comment]
> Hello,
> 
> Thank you for reaching out and providing detailed information about your issue. It's great to see your proactive approach in integrating YOLOv8 OBB with NCNN for your mobile application. Let's work through this together.
> 
> ### Steps to Troubleshoot and Resolve the Issue
> 1. **Verify Model Conversion**:
>    Ensure that the model conversion process is correct. The command you used seems appropriate, but let's confirm that the latest versions of `torch` and `ultralytics` are installed. You can update them using:
>    ```shell
>    pip install --upgrade torch ultralytics
>    ```
> 2. **Check Output Shapes**:
>    The discrepancy in the output shapes is a key issue. The output shape should be consistent with the expected dimensions for YOLOv8 OBB. The reshaping you applied in C++ might not be sufficient. Instead, let's ensure the model's output is correctly interpreted.
> 3. **Model Testing in Python**:
>    Before integrating the model into your C++ application, let's verify its correctness in Python. Here’s a sample code snippet to test the exported NCNN model:
>    ```python
>    import ncnn
>    from ultralytics import YOLO
>    
>    # Load the model
>    model = YOLO("yolov8n-obb.pt")
>    
>    # Export to NCNN
>    model.export(format="ncnn", opset=12)
>    
>    # Load the exported NCNN model
>    net = ncnn.Net()
>    net.load_param("yolov8n-obb_ncnn_model.param")
>    net.load_model("yolov8n-obb_ncnn_model.bin")
>    
>    # Prepare input
>    mat_in = ncnn.Mat.from_pixels_resize(image, ncnn.Mat.PixelType.PIXEL_BGR, width, height, target_width, target_height)
>    
>    # Run inference
>    extractor = net.create_extractor()
>    extractor.input("data", mat_in)
>    ret, mat_out = extractor.extract("output")
>    
>    print(mat_out)
>    ```
> 4. **Adjusting the Output in C++**:
>    If the Python test works correctly, ensure that the C++ code correctly interprets the model's output. The reshaping logic should align with the model's expected output dimensions. Here’s a refined approach:
>    ```c++
>    if (out.w > out.h) {
>        out = out.reshape(out.h, out.w);
>    } else {
>        // Ensure the reshaping logic aligns with the model's expected output
>        // Adjust according to the specific output dimensions of your model
>    }
>    ```
> 5. **Bounding Box Accuracy**:
>    Ensure that the post-processing steps for bounding box extraction are correctly implemented. This includes decoding the bounding box coordinates and applying any necessary transformations.
> 
> ### Additional Resources
> For further guidance, you might find these resources helpful:
> 
> * [Ultralytics YOLOv8 Documentation](https://docs.ultralytics.com)
> * [NCNN Documentation](https://github.com/Tencent/ncnn/wiki)
> 
> If the issue persists, please provide a minimum reproducible example of your code, including the specific steps and configurations used. This will help us better understand the problem and provide more targeted assistance. You can refer to our guide on creating a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example).
> 
> Feel free to reach out with any further questions or updates on your progress. We're here to help!

Thank you for your quick response and the troubleshooting steps. I followed the steps, but the issue still exists. It might be related to the post-process, as you mentioned. Can you confirm if the output from "NCNN OBB" is correct? Specifically, does "H=20" mean:

[0-3] correspond to x, y, w, h
[4-18] correspond to class scores (15)
[19] is the angle
Thank you for your help.

**Comment by siri-sinthawat** [comment]
@zgsxwsdxg I'm still stuck on the post-process. If I solve it, I will update you as soon as possible.

**Comment by siri-sinthawat** [comment]
I'm having trouble understanding the output differences when running YOLOv8-obb.pt with the **non_max_suppression** function. When I run the model directly, 
```
from ultralytics import YOLO

model = YOLO("yolov8n-obb.pt", task='obb')
results = model.predict('bus.jpg')
```
I get one result box, 
<img width="596" alt="Screenshot 2024-06-13 at 16 51 44" src="https://github.com/ultralytics/ultralytics/assets/49611395/9fdf1d3c-b55a-4c58-95e5-3f4e5cfa5ae9">

but after converting it to an NCNN model and performing inference on the same image, 
```
from ultralytics import YOLO
import ncnn

model = YOLO("yolov8n-obb.pt")
model.export(format="ncnn", opset=12)
```
```
net = ncnn.Net()
net.load_param("yolov8n-obb_ncnn_model/model.ncnn.param")
net.load_model("yolov8n-obb_ncnn_model/model.ncnn.bin")

image = cv2.imread('bus.jpg')
mat_in = ncnn.Mat.from_pixels_resize(image, ncnn.Mat.PixelType.PIXEL_BGR, image.shape[0], image.shape[1],
                                         1024,024)

extractor = net.create_extractor()
extractor.input("in0", mat_in)
ret, mat_out = extractor.extract("out0")
```
I get a ncnn::Mat. I then converted this to a tensor and used the non_max_suppression function from the Ultralytics library,
```
import numpy as np
from ultralytics.utils.ops import non_max_suppression

mat_out = np.array(mat_out)
mat_out = mat2tensor(mat_out.data, mat_out.shape[0], mat_out.shape[1])

preds = non_max_suppression(mat_out, nc=15, rotated=True,
                                iou_thres=0.7)
```
but now I'm getting 300 result boxes. 
<img width="713" alt="Screenshot 2024-06-13 at 17 03 43" src="https://github.com/ultralytics/ultralytics/assets/49611395/ec0e221d-2dd8-4783-ade5-3e1434f8070c">

Can @glenn-jocher suggest what might be causing this discrepancy? I'm quite confused about it.

**Comment by glenn-jocher** [comment]
Hello @siri-sinthawat,

Thank you for providing detailed information about the issue you're facing. Let's work through this step-by-step to identify the root cause and find a solution.

### Steps to Troubleshoot and Resolve the Issue

1. **Verify Model Conversion**:
   Ensure that the model conversion process is correct. The command you used seems appropriate, but let's confirm that the latest versions of `torch` and `ultralytics` are installed. You can update them using:
   ```bash
   pip install --upgrade torch ultralytics
   ```

2. **Check Output Shapes**:
   The discrepancy in the output shapes is a key issue. The output shape should be consistent with the expected dimensions for YOLOv8 OBB. The reshaping you applied in C++ might not be sufficient. Instead, let's ensure the model's output is correctly interpreted.

3. **Model Testing in Python**:
   Before integrating the model into your C++ application, let's verify its correctness in Python. Here’s a sample code snippet to test the exported NCNN model:
   ```python
   import ncnn
   from ultralytics import YOLO

   # Load the model
   model = YOLO("yolov8n-obb.pt")

   # Export to NCNN
   model.export(format="ncnn", opset=12)

   # Load the exported NCNN model
   net = ncnn.Net()
   net.load_param("yolov8n-obb_ncnn_model.param")
   net.load_model("yolov8n-obb_ncnn_model.bin")

   # Prepare input
   mat_in = ncnn.Mat.from_pixels_resize(image, ncnn.Mat.PixelType.PIXEL_BGR, width, height, target_width, target_height)

   # Run inference
   extractor = net.create_extractor()
   extractor.input("data", mat_in)
   ret, mat_out = extractor.extract("output")

   print(mat_out)
   ```

4. **Adjusting the Output in C++**:
   If the Python test works correctly, ensure that the C++ code correctly interprets the model's output. The reshaping logic should align with the model's expected output dimensions. Here’s a refined approach:
   ```c++
   if (out.w > out.h) {
       out = out.reshape(out.h, out.w);
   } else {
       // Ensure the reshaping logic aligns with the model's expected output
       // Adjust according to the specific output dimensions of your model
   }
   ```

5. **Bounding Box Accuracy**:
   Ensure that the post-processing steps for bounding box extraction are correctly implemented. This includes decoding the bounding box coordinates and applying any necessary transformations.

### Additional Resources

For further guidance, you might find these resources helpful:
- [Ultralytics YOLOv8 Documentation](https://docs.ultralytics.com)
- [NCNN Documentation](https://github.com/Tencent/ncnn/wiki)

If the issue persists, please provide a minimum reproducible example of your code, including the specific steps and configurations used. This will help us better understand the problem and provide more targeted assistance. You can refer to our guide on creating a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example).

Feel free to reach out with any further questions or updates on your progress. We're here to help!

**Comment by siri-sinthawat** [comment]
Good day @glenn-jocher,

Thank you for your prompt support and suggestions. I'm currently experiencing some discrepancies in the output when running the YOLOv8-obb.pt model compared to the version I converted to NCNN. I used the **`non_max_suppression`** function from Ultralytics to ensure that both model types (.pt, .bin/.param) produce consistent results for the same image.

### Scenario 1: Direct Model Execution
Using the model directly:

```python
from ultralytics import YOLO

# Load the model
model = YOLO("yolov8n-obb.pt", task='obb')
# Run predict
results = model.predict('bus.jpg')
```
I get one result box.
<img width="651" alt="Screenshot 2024-06-14 at 8 23 23" src="https://github.com/ultralytics/ultralytics/assets/49611395/d46fe6db-f6ad-4eee-adae-40baba607002">


### Scenario 2: NCNN Model Conversion and Inference
After converting the model to NCNN and performing inference on the same image:

```python
from ultralytics import YOLO
import ncnn
import cv2
import numpy as np
import torch
from ultralytics.utils.ops import non_max_suppression

# Load the model
model = YOLO("yolov8n-obb.pt")
# Export to NCNN
model.export(format="ncnn", opset=12)

# Load the exported NCNN model
net = ncnn.Net()
net.load_param("yolov8n-obb_ncnn_model/model.ncnn.param")
net.load_model("yolov8n-obb_ncnn_model/model.ncnn.bin")

# Prepare input
image = cv2.imread('bus.jpg')
mat_in = ncnn.Mat.from_pixels_resize(image, ncnn.Mat.PixelType.PIXEL_BGR, image.shape[0], image.shape[1], 1024, 1024)

# Run inference
extractor = net.create_extractor()
extractor.input("in0", mat_in)
ret, mat_out = extractor.extract("out0")

# Convert NCNN output to tensor and apply non-max suppression
mat_out = np.array(mat_out)
np_array = np.array(mat_out.data, dtype=np.float32).reshape((mat_out.shape[0], mat_out.shape[1]))
tensor_data = torch.from_numpy(np_array)
tensor_data = tensor_data.unsqueeze(0)  # Add a singleton dimension to match the desired shape (1, 20, 21504)

# Run predict
result = non_max_suppression(tensor_data, nc=15, rotated=True, iou_thres=0.7)
```
This yields 300 result boxes.
<img width="745" alt="Screenshot 2024-06-14 at 8 54 18" src="https://github.com/ultralytics/ultralytics/assets/49611395/da37f9ce-7af6-41b9-9185-54894dd8b007">


### Observations
The NCNN model yields very high scores for all labels, often between 0.99998 and 1.0, which is significantly higher than the scores from the original .pt model.

### Request for Help
Can anyone suggest what might be causing this discrepancy? I'm quite confused about the differences in scores and the number of result boxes.

**Comment by glenn-jocher** [comment]
Hello @siri-sinthawat,

Thank you for providing detailed information and the code snippets. It's great to see your proactive approach in troubleshooting the issue. Let's work through this together to identify the root cause and find a solution.

### Key Points to Address

1. **Model Conversion Verification**:
   Ensure that the model conversion process from PyTorch to NCNN is correct. The command you used seems appropriate, but let's confirm that the latest versions of `torch` and `ultralytics` are installed. You can update them using:
   ```bash
   pip install --upgrade torch ultralytics
   ```

2. **Output Shape and Post-Processing**:
   The discrepancy in the number of result boxes and the high scores indicates a potential issue with the output shape or the post-processing steps. Let's ensure that the output from the NCNN model is correctly interpreted and processed.

3. **Non-Max Suppression (NMS)**:
   The high number of result boxes suggests that the NMS might not be functioning as expected. Ensure that the input to the NMS function is correctly formatted.

### Steps to Troubleshoot

1. **Verify Output Shape**:
   Ensure that the output shape from the NCNN model matches the expected dimensions. The reshaping logic should align with the model's expected output dimensions.

2. **Check Post-Processing**:
   Ensure that the post-processing steps, including the conversion of the NCNN output to a tensor and the application of NMS, are correctly implemented.

### Example Code for Post-Processing

Here’s a refined approach to ensure the output is correctly processed:

```python
import ncnn
import cv2
import numpy as np
import torch
from ultralytics.utils.ops import non_max_suppression

# Load the model
model = YOLO("yolov8n-obb.pt")
# Export to NCNN
model.export(format="ncnn", opset=12)

# Load the exported NCNN model
net = ncnn.Net()
net.load_param("yolov8n-obb_ncnn_model/model.ncnn.param")
net.load_model("yolov8n-obb_ncnn_model/model.ncnn.bin")

# Prepare input
image = cv2.imread('bus.jpg')
mat_in = ncnn.Mat.from_pixels_resize(image, ncnn.Mat.PixelType.PIXEL_BGR, image.shape[1], image.shape[0], 1024, 1024)

# Run inference
extractor = net.create_extractor()
extractor.input("in0", mat_in)
ret, mat_out = extractor.extract("out0")

# Convert NCNN output to tensor and apply non-max suppression
mat_out = np.array(mat_out)
np_array = np.array(mat_out.data, dtype=np.float32).reshape((mat_out.shape[0], mat_out.shape[1]))
tensor_data = torch.from_numpy(np_array)
tensor_data = tensor_data.unsqueeze(0)  # Add a singleton dimension to match the desired shape (1, 20, 21504)

# Run non-max suppression
result = non_max_suppression(tensor_data, nc=15, rotated=True, iou_thres=0.7)

# Print the results
print(result)
```

### Additional Considerations

- **Score Threshold**: Ensure that the score threshold for NMS is set appropriately. High scores might indicate that the threshold is too low.
- **Debugging**: Print intermediate outputs to verify that each step is producing the expected results.

### Community and Support

If the issue persists, please provide a minimum reproducible example of your code, including the specific steps and configurations used. This will help us better understand the problem and provide more targeted assistance. You can refer to our guide on creating a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example).

Feel free to reach out with any further questions or updates on your progress. We're here to help! 😊

**Comment by siri-sinthawat** [comment]
Good day @glenn-jocher,

Thank you for providing the troubleshooting steps. The results have improved, but they are still not quite right. I found a helpful [repository](https://github.com/ZeonlungPun/YoloDeploymentWithOnnxruntimeAndOpencv/blob/main/yolov8obbOPENCV/yolov8obb.cpp) that seems to solve a similar problem.

I'm trying to convert and run a function from C++ to Python to post-process the output of an ONNX and NCNN. The ONNX model inference result works correctly, but the NCNN inference result gives incorrect results despite having the same output shape and post-processing steps.

### Environment
Operating System: macOS 14.5
Python Version:  3.9.19
Ultralytics Version: 8.2.34
NCNN Version: 1.0.20240410

### ONNX Inference and Post-Processing
Here is the Python code for the ONNX inference and post-processing:
```Python
import math

import cv2
import numpy as np
from onnxruntime import InferenceSession
from ultralytics import YOLO

pi = math.pi  # Use math.pi for clarity
def preprocess(img, size: int = 1024):
    img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
    img = cv2.resize(img, (size, size))
    img = img.astype(np.float32)
    img = img / 255.0  # Normalize to [0, 1]
    img = np.transpose(img, (2, 0, 1))  # Change data layout from HWC to CHW
    img = np.expand_dims(img, axis=0)  # Add batch dimension
    return img

def nms(outputs, img, score_threshold=0.7, nms_threshold=0.25, classes: int = 15):
    rows = outputs.shape[2]
    dimensions = outputs.shape[1]

    outputs = outputs.reshape(dimensions, -1).transpose()
    data = outputs.flatten()

    x_factor = img.shape[1] / 1024.0
    y_factor = img.shape[0] / 1024.0

    class_ids = []
    confidences = []
    boxes = []

    for i in range(rows):
        classes_scores = data[4:4 + classes]
        maxClassScore = np.max(classes_scores)
        class_id = np.argmax(classes_scores)

        if maxClassScore > score_threshold:
            confidences.append(maxClassScore)
            class_ids.append(class_id)

            x = data[0] * x_factor
            y = data[1] * y_factor
            w = data[2] * x_factor
            h = data[3] * y_factor

            angle = data[dimensions - 1]
            if 0.5 * pi <= angle <= 0.75 * pi:
                angle -= pi

            box = cv2.RotatedRect((x, y), (w, h), angle * 180 / pi)
            boxes.append(box)

        data = data[dimensions:]

    indices = cv2.dnn.NMSBoxesRotated(
        boxes,
        confidences,
        score_threshold,
        nms_threshold
    )
    return [boxes[i] for i in range(len(boxes)) if i in indices]

def draw(image, detect_boxes):
    for box in detect_boxes:
        points = cv2.boxPoints(box)
        points = np.intp(points)
        for i in range(4):
            cv2.line(image, tuple(points[i]), tuple(points[(i + 1) % 4]), (0, 0, 255), 2)

    cv2.imshow("RotatedRect", image)
    cv2.waitKey(0)

def main():    
    # Load the model and export to ONNX
    model = YOLO("yolov8n-obb.pt")
    model.export(format="onnx")

    # Load the ONNX model
    onnx_model_path = "yolov8n-obb.onnx"
    ort_session = InferenceSession(onnx_model_path)

    # Read and preprocess the input image
    img = cv2.imread("obb.jpeg")
    input_tensor = preprocess(img)

    # Run inference
    ort_inputs = {ort_session.get_inputs()[0].name: input_tensor}
    outputs = ort_session.run(None, ort_inputs)[0]

    # Post-process the outputs and draw the detections
    detected_boxes = nms(outputs, img, score_threshold=0.07)
    draw(img, detected_boxes)


if __name__ == '__main__':
    main()

```
### Output
![onnx_result](https://github.com/ultralytics/ultralytics/assets/49611395/649bfa1d-507f-42d7-9473-e154df377c74)


### NCNN Inference and Post-Processing
Here is the Python code for the NCNN inference and post-processing:
```Python
import math

import cv2
import numpy as np
from ncnn import ncnn
from ultralytics import YOLO

pi = math.pi  # Use math.pi for clarity
def preprocess(img, size: int = 1024):
    img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
    img = cv2.resize(img, (size, size))
    img = img.astype(np.float32)
    img = img / 255.0  # Normalize to [0, 1]
    img = np.transpose(img, (2, 0, 1))  # Change data layout from HWC to CHW
    img = np.expand_dims(img, axis=0)  # Add batch dimension
    return img

def nms(outputs, img, score_threshold=0.7, nms_threshold=0.25, classes: int = 15):
    rows = outputs.shape[2]
    dimensions = outputs.shape[1]

    outputs = outputs.reshape(dimensions, -1).transpose()
    data = outputs.flatten()

    x_factor = img.shape[1] / 1024.0
    y_factor = img.shape[0] / 1024.0

    class_ids = []
    confidences = []
    boxes = []

    for i in range(rows):
        classes_scores = data[4:4 + classes]
        maxClassScore = np.max(classes_scores)
        class_id = np.argmax(classes_scores)

        if maxClassScore > score_threshold:
            confidences.append(maxClassScore)
            class_ids.append(class_id)

            x = data[0] * x_factor
            y = data[1] * y_factor
            w = data[2] * x_factor
            h = data[3] * y_factor

            angle = data[dimensions - 1]
            if 0.5 * pi <= angle <= 0.75 * pi:
                angle -= pi

            box = cv2.RotatedRect((x, y), (w, h), angle * 180 / pi)
            boxes.append(box)

        data = data[dimensions:]

    indices = cv2.dnn.NMSBoxesRotated(
        boxes,
        confidences,
        score_threshold,
        nms_threshold
    )
    return [boxes[i] for i in range(len(boxes)) if i in indices]

def draw(image, detect_boxes):
    for box in detect_boxes:
        points = cv2.boxPoints(box)
        points = np.intp(points)
        for i in range(4):
            cv2.line(image, tuple(points[i]), tuple(points[(i + 1) % 4]), (0, 0, 255), 2)

    cv2.imshow("RotatedRect", image)
    cv2.waitKey(0)

def main():
    # Load the model and export to NCNN
    model = YOLO("yolov8n-obb.pt")
    model.export(format="ncnn", opset=12)

    # Load the ONNX model
    net = ncnn.Net()
    net.load_param("yolov8n-obb_ncnn_model/model.ncnn.param")
    net.load_model("yolov8n-obb_ncnn_model/model.ncnn.bin")

    # Read and preprocess the input image
    img = cv2.imread("obb.jpeg")
    mat_in = ncnn.Mat.from_pixels_resize(img, ncnn.Mat.PixelType.PIXEL_RGB, img.shape[1], img.shape[0], 1024, 1024)

    # Run inference
    extractor = net.create_extractor()
    extractor.input("in0", mat_in)
    ret, mat_out = extractor.extract("out0")

    # Convert NCNN output to tensor and apply non-max suppression
    mat_out = np.array(mat_out)
    outputs = np.array(mat_out.data, dtype=np.float32).reshape((1, mat_out.shape[0], mat_out.shape[1]))

    # Post-process the outputs and draw the detections
    detected_boxes = nms(outputs, img, score_threshold=0.07)
    draw(img, detected_boxes)


if __name__ == '__main__':
    main()

```
### Output
![ncnn_result](https://github.com/ultralytics/ultralytics/assets/49611395/427fe58f-ead1-475d-bb39-735a94d4b122)

### Observations
The NCNN model yields very high scores for all labels, often between 0.99998 and 1.0, which is significantly higher than the scores from the .ONNX model.

### Request for Help
Can anyone suggest what might be causing this discrepancy? I'm quite confused about the differences in scores and the number of result boxes.

**Comment by glenn-jocher** [comment]
Good day @siri-sinthawat,

Thank you for providing detailed information and code snippets. It's great to see your proactive approach in troubleshooting the issue. Let's work through this together to identify the root cause and find a solution.

### Key Points to Address

1. **Model Conversion Verification**:
   Ensure that the model conversion process from PyTorch to NCNN is correct. The command you used seems appropriate, but let's confirm that the latest versions of `torch` and `ultralytics` are installed. You can update them using:
   ```bash
   pip install --upgrade torch ultralytics
   ```

2. **Output Shape and Post-Processing**:
   The discrepancy in the number of result boxes and the high scores indicates a potential issue with the output shape or the post-processing steps. Let's ensure that the output from the NCNN model is correctly interpreted and processed.

3. **Non-Max Suppression (NMS)**:
   The high number of result boxes suggests that the NMS might not be functioning as expected. Ensure that the input to the NMS function is correctly formatted.

### Steps to Troubleshoot

1. **Verify Output Shape**:
   Ensure that the output shape from the NCNN model matches the expected dimensions. The reshaping logic should align with the model's expected output dimensions.

2. **Check Post-Processing**:
   Ensure that the post-processing steps, including the conversion of the NCNN output to a tensor and the application of NMS, are correctly implemented.

### Example Code for Post-Processing

Here’s a refined approach to ensure the output is correctly processed:

```python
import ncnn
import cv2
import numpy as np
import torch
from ultralytics.utils.ops import non_max_suppression

# Load the model
model = YOLO("yolov8n-obb.pt")
# Export to NCNN
model.export(format="ncnn", opset=12)

# Load the exported NCNN model
net = ncnn.Net()
net.load_param("yolov8n-obb_ncnn_model/model.ncnn.param")
net.load_model("yolov8n-obb_ncnn_model/model.ncnn.bin")

# Prepare input
image = cv2.imread('bus.jpg')
mat_in = ncnn.Mat.from_pixels_resize(image, ncnn.Mat.PixelType.PIXEL_BGR, image.shape[1], image.shape[0], 1024, 1024)

# Run inference
extractor = net.create_extractor()
extractor.input("in0", mat_in)
ret, mat_out = extractor.extract("out0")

# Convert NCNN output to tensor and apply non-max suppression
mat_out = np.array(mat_out)
np_array = np.array(mat_out.data, dtype=np.float32).reshape((mat_out.shape[0], mat_out.shape[1]))
tensor_data = torch.from_numpy(np_array)
tensor_data = tensor_data.unsqueeze(0)  # Add a singleton dimension to match the desired shape (1, 20, 21504)

# Run non-max suppression
result = non_max_suppression(tensor_data, nc=15, rotated=True, iou_thres=0.7)

# Print the results
print(result)
```

### Additional Considerations

- **Score Threshold**: Ensure that the score threshold for NMS is set appropriately. High scores might indicate that the threshold is too low.
- **Debugging**: Print intermediate outputs to verify that each step is producing the expected results.

### Community and Support

If the issue persists, please provide a minimum reproducible example of your code, including the specific steps and configurations used. This will help us better understand the problem and provide more targeted assistance. You can refer to our guide on creating a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example).

Feel free to reach out with any further questions or updates on your progress. We're here to help! 😊

**Comment by siri-sinthawat** [comment]
Good day everyone!

I wanted to share that I found a solution that works for me in the NCNN community. I encountered an issue in my code where I incorrectly handled input and skipped some processes. By using the following code before sending the Mat input for inference, I was able to get the correct result with the right score threshold:

```Python
import cv2
import numpy as np
from ncnn import ncnn
from ultralytics import YOLO


# Load the model
model = YOLO("yolov8n-obb.pt")
# Export to NCNN
model.export(format="ncnn")

# Load the exported NCNN model
net = ncnn.Net()
net.load_param("yolov8n-obb_ncnn_model/model.ncnn.param")
net.load_model("yolov8n-obb_ncnn_model/model.ncnn.bin")

target_size = 1024
MAX_STRIDE = 32

# Prepare input
image = cv2.imread('obb.jpeg')
img_h, img_w, _ = image.shape

# Letterbox pad to multiple of MAX_STRIDE
w = img_w
h = img_h
scale = 1.0
if w > h:
    scale = target_size / w
    w = target_size
    h = int(h * scale)
else:
    scale = target_size / h
    h = target_size
    w = int(w * scale)

mat_in = ncnn.Mat.from_pixels_resize(image.data, ncnn.Mat.PixelType.PIXEL_BGR2RGB, img_w, img_h,
                                     w, h)

wpad = ((target_size + MAX_STRIDE - 1) // MAX_STRIDE) * MAX_STRIDE - w
hpad = ((target_size + MAX_STRIDE - 1) // MAX_STRIDE) * MAX_STRIDE - h

in_pad = ncnn.copy_make_border(
    mat_in,
    hpad // 2,
    hpad - hpad // 2,
    wpad // 2,
    wpad - wpad // 2,
    ncnn.BorderType.BORDER_CONSTANT,
    114,
)

norm_vals = np.array([1 / 255.0, 1 / 255.0, 1 / 255.0], dtype=np.float32)
in_pad.substract_mean_normalize([0, 0, 0], norm_vals)

# Run inference
extractor = net.create_extractor()
extractor.input("in0", in_pad)
ret, mat_out = extractor.extract("out0")

# Print the results
result = np.array(mat_out)
print(result)
```
this is my output:
![obb_result](https://github.com/ultralytics/ultralytics/assets/49611395/ab5b20e1-6a93-4249-a591-947cbb1aea83)


This adjustment made a significant difference in the outcome. I hope this information can be helpful to others facing a similar challenge.

**Comment by HarinathPapireddy** [comment]
How to draw the bounding boxes from the result can you share the code

**Comment by glenn-jocher** [comment]
Certainly! You can use the `cv2.rectangle` function from OpenCV to draw bounding boxes on your images. If you need further assistance, please let me know!

**Comment by siri-sinthawat** [comment]
@HarinathPapireddy To draw bounding boxes from the NCNN model results, I first convert the model's output into cv2.RotatedRect format. Once the conversion is done, I use the following code to draw these bounding boxes on the image:

```Python
import cv2
import numpy as np

def paint(image: cv2.Mat, boxes: list[cv2.RotatedRect], color=(0, 0, 255)):
    for box in boxes:
        # Get the four vertices of the rotated rectangle
        vertices = cv2.boxPoints(box)
        vertices = np.intp(vertices)

        # Draw the contours of the rotated rectangle
        cv2.drawContours(image, [vertices], 0, color, 2)

    # Display the image
    cv2.imshow('Annotated Image', image)
    cv2.waitKey(0)
    cv2.destroyAllWindows()
```

**Comment by HarinathPapireddy** [comment]
Thank you @siri-sinthawat

**Comment by gangwu001** [comment]
> Good day everyone!
> 
> I wanted to share that I found a solution that works for me in the NCNN community. I encountered an issue in my code where I incorrectly handled input and skipped some processes. By using the following code before sending the Mat input for inference, I was able to get the correct result with the right score threshold:
> 
> import cv2
> import numpy as np
> from ncnn import ncnn
> from ultralytics import YOLO
> 
> 
> # Load the model
> model = YOLO("yolov8n-obb.pt")
> # Export to NCNN
> model.export(format="ncnn")
> 
> # Load the exported NCNN model
> net = ncnn.Net()
> net.load_param("yolov8n-obb_ncnn_model/model.ncnn.param")
> net.load_model("yolov8n-obb_ncnn_model/model.ncnn.bin")
> 
> target_size = 1024
> MAX_STRIDE = 32
> 
> # Prepare input
> image = cv2.imread('obb.jpeg')
> img_h, img_w, _ = image.shape
> 
> # Letterbox pad to multiple of MAX_STRIDE
> w = img_w
> h = img_h
> scale = 1.0
> if w > h:
>     scale = target_size / w
>     w = target_size
>     h = int(h * scale)
> else:
>     scale = target_size / h
>     h = target_size
>     w = int(w * scale)
> 
> mat_in = ncnn.Mat.from_pixels_resize(image.data, ncnn.Mat.PixelType.PIXEL_BGR2RGB, img_w, img_h,
>                                      w, h)
> 
> wpad = ((target_size + MAX_STRIDE - 1) // MAX_STRIDE) * MAX_STRIDE - w
> hpad = ((target_size + MAX_STRIDE - 1) // MAX_STRIDE) * MAX_STRIDE - h
> 
> in_pad = ncnn.copy_make_border(
>     mat_in,
>     hpad // 2,
>     hpad - hpad // 2,
>     wpad // 2,
>     wpad - wpad // 2,
>     ncnn.BorderType.BORDER_CONSTANT,
>     114,
> )
> 
> norm_vals = np.array([1 / 255.0, 1 / 255.0, 1 / 255.0], dtype=np.float32)
> in_pad.substract_mean_normalize([0, 0, 0], norm_vals)
> 
> # Run inference
> extractor = net.create_extractor()
> extractor.input("in0", in_pad)
> ret, mat_out = extractor.extract("out0")
> 
> # Print the results
> result = np.array(mat_out)
> print(result)
> Here is my output: ![obb_result](https://github.com/ultralytics/ultralytics/assets/49611395/ab5b20e1-6a93-4249-a591-947cbb1aea83)
> 
> This adjustment made a significant difference in the outcome. I hope this information can be helpful to others facing a similar challenge.

Can you provide the complete code? Thank you very much!

**Comment by glenn-jocher** [comment]
@gangwu001 Here’s a minimal end-to-end example that exports an OBB model to NCNN, runs inference by loading the exported NCNN folder directly with Ultralytics, and draws rotated boxes; replace the `model=` path with your weights as needed:

```python
import cv2
import numpy as np
from ultralytics import YOLO

# 1) Export to NCNN (OBB)
YOLO("yolov8n-obb.pt").export(format="ncnn", imgsz=1024)  # creates 'yolov8n-obb_ncnn_model/'

# 2) Load exported NCNN model with Ultralytics runtime
model = YOLO("yolov8n-obb_ncnn_model")  # or your custom *_ncnn_model directory

# 3) Inference
results = model("obb.jpeg", imgsz=1024)

# 4) Draw OBBs
im = results[0].orig_img.copy()
obb = results[0].obb  # oriented boxes
if obb is not None:
    quads = obb.xyxyxyxy.astype(int)  # (N, 4, 2)
    for q in quads:
        cv2.polylines(im, [q], True, (0, 0, 255), 2)
cv2.imwrite("vis_obb.jpg", im)
```

For format details and alternatives (including direct NCNN usage), see the NCNN export guide at [How to Export to NCNN](https://docs.ultralytics.com/integrations/ncnn/) and OBB specifics at [Oriented Bounding Boxes (OBB)](https://docs.ultralytics.com/tasks/obb/). If you still need a pure-NCNN example with manual letterbox/pad and decoding, please open a new issue with a small reproducible snippet so the team can help.

**Comment by gangwu001** [comment]
What does the format or fields of mat_out in ret, mat_out = extractor.extract("out0") mean? I'm using NCNN to implement PIG-based OBB detection (only one category). I've followed the tutorial to convert the Yolo11n-obb to ncnn format and loaded the param and bin files into ncnn.Net(). The shape of mat_out is [6, 21504]. Does the 6 represent the information [x, y, w, h, θ, obj_conf, class_con

<img width="2526" height="414" alt="Image" src="https://github.com/user-attachments/assets/d2a59a9a-759e-4007-bc2f-2dc22405449f" />

**Comment by glenn-jocher** [comment]
Hi @gangwu001 — you’re close. For a single-class YOLO OBB NCNN export, out0 is shaped [no, anchors]. With imgsz=1024 the anchor count is 21504, and no = 4 + nc + 1, i.e. [cx, cy, w, h, class_score, angle]. There is no separate objectness channel. So for your [6, 21504] tensor: first 4 = box, 5th = class score, 6th = angle (radians). After extraction, transpose to (21504, 6), threshold on the class score, then run rotated NMS using the angle. See the NCNN export guide and the class/extra split in NMSModel for details: [How to Export to NCNN](https://docs.ultralytics.com/integrations/ncnn/) and [Exporter reference (NMSModel splitting boxes/scores/extras)](https://docs.ultralytics.com/reference/engine/exporter/); OBB result fields are described in the [OBB task docs](https://docs.ultralytics.com/tasks/obb/).

**Comment by makanov001** [comment]
> [@HarinathPapireddy](https://github.com/HarinathPapireddy) To draw bounding boxes from the OBB NCNN model results, I first convert the model's output into cv2.RotatedRect format. Once the conversion is done, I use the following code to draw these bounding boxes on the image:
> 
> import cv2
> import numpy as np
> 
> def paint(image: cv2.Mat, boxes: list[cv2.RotatedRect], color=(0, 0, 255)):
>     for box in boxes:
>         # Get the four vertices of the rotated rectangle
>         vertices = cv2.boxPoints(box)
>         vertices = np.intp(vertices)
> 
>         # Draw the contours of the rotated rectangle
>         cv2.drawContours(image, [vertices], 0, color, 2)
> 
>     # Display the image
>     cv2.imshow('Image', image)
>     cv2.waitKey(0)
>     cv2.destroyAllWindows()

Could you please tell me if you have performed NMS on the "output"? Also, could you provide the subsequent operation codes for the "output" (mat_out), namely sorting, NMS, and image drawing?

**Comment by glenn-jocher** [comment]
Yes — `mat_out` is the raw head output, so you still need confidence filtering, class-wise rotated NMS, and then drawing; for single-class OBB each row is `[cx, cy, w, h, score, angle]`, and for multi-class it is `[cx, cy, w, h, cls0 ... clsN-1, angle]` with no separate objectness channel.

```python
import cv2
import numpy as np

def postprocess_obb(mat_out, scale, pad_w, pad_h, nc=1, conf=0.25, iou=0.45):
    pred = np.array(mat_out, dtype=np.float32)
    if pred.ndim != 2:
        pred = pred.reshape(pred.shape[0], -1)
    if pred.shape[0] < pred.shape[1]:
        pred = pred.T  # -> (num_anchors, 4 + nc + 1)

    dets = []
    boxes_by_cls, scores_by_cls = {}, {}

    for row in pred:
        cls_scores = row[4:4 + nc]
        cls_id = int(np.argmax(cls_scores))
        score = float(cls_scores[cls_id])
        if score < conf:
            continue

        cx, cy, w, h = row[:4]
        angle = float(row[4 + nc])  # radians

        # undo letterbox padding + resize
        cx = (cx - pad_w) / scale
        cy = (cy - pad_h) / scale
        w /= scale
        h /= scale

        box = ((cx, cy), (w, h), angle * 180.0 / np.pi)
        boxes_by_cls.setdefault(cls_id, []).append(box)
        scores_by_cls.setdefault(cls_id, []).append(score)

    for cls_id, boxes in boxes_by_cls.items():
        keep = cv2.dnn.NMSBoxesRotated(boxes, scores_by_cls[cls_id], conf, iou)
        keep = np.array(keep).reshape(-1) if len(keep) else []
        for i in keep:
            dets.append((cls_id, scores_by_cls[cls_id][i], boxes[i]))

    dets.sort(key=lambda x: x[1], reverse=True)
    return dets


def draw_obb(image, dets, color=(0, 0, 255), thickness=2):
    im = image.copy()
    for cls_id, score, box in dets:
        pts = cv2.boxPoints(box).astype(np.int32)
        cv2.polylines(im, [pts], True, color, thickness)
        cv2.putText(im, f"{cls_id}:{score:.2f}", tuple(pts[0]),
                    cv2.FONT_HERSHEY_SIMPLEX, 0.6, color, 2)
    return im


# use the same `scale`, `wpad`, `hpad` values from your preprocessing step
dets = postprocess_obb(mat_out, scale=scale, pad_w=wpad / 2, pad_h=hpad / 2, nc=1, conf=0.25, iou=0.45)
vis = draw_obb(image, dets)
cv2.imwrite("obb_vis.jpg", vis)
```

If you want behavior aligned with Ultralytics postprocessing, the closest reference is the [NMS reference](https://docs.ultralytics.com/reference/utils/nms).
